babel/numbers.py

```diff
@@ -641,68 +698,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
         self.frac_prec = frac_prec
         self.exp_prec = exp_prec
         self.exp_plus = exp_plus
-        if '%' in ''.join(self.prefix + self.suffix):
-            self.scale = 2
-        elif u'‰' in ''.join(self.prefix + self.suffix):
-            self.scale = 3
-        else:
-            self.scale = 0
 
     def __repr__(self):
         return '<%s %r>' % (type(self).__name__, self.pattern)
 
-    def apply(self, value, locale, currency=None, force_frac=None):
-        frac_prec = force_frac or self.frac_prec
+    @property
+    def scale(self):
+        """Scaling factor to apply to the number before rendering.
+
+        Auto-set to a factor of 2 or 3 if presence of a % or ‰ sign is
+        detected in the pattern. Default is to not mess with the scale at all.
+        """
+        scale = 0
+        if '%' in ''.join(self.prefix + self.suffix):
+            scale = 2
+        elif u'‰' in ''.join(self.prefix + self.suffix):
+            scale = 3
+        return scale
+
+    @staticmethod
+    def precision(number):
+        """Return maximum precision of a decimal instance's fractional part.
+
+        Precision is extracted from the fractional part only.
+        """
+        # Copied from: https://github.com/mahmoud/boltons/pull/59
+        decimal_tuple = number.normalize().as_tuple()
+        if decimal_tuple.exponent >= 0:
+            return 0
+        return abs(decimal_tuple.exponent)
+
+    @staticmethod
+    def quantum(precision):
+        """Return minimal quantum of a number, as defined by precision."""
+        return decimal.Decimal(10) ** (-precision)
+
+    def apply(
+            self, value, locale, currency=None, currency_digits=True,
+            decimal_quantization=True):
+        """Renders into a string a number following the defined pattern.
+
+        Forced decimal quantization is active by default so we'll produce a
+        number string that is strictly following CLDR pattern definitions.
+        """
+        # Manipulates decimal instances only.
         if not isinstance(value, decimal.Decimal):
             value = decimal.Decimal(str(value))
+
+        # Scale the number.
         value = value.scaleb(self.scale)
+
+        # Separate the absolute value from its sign.
         is_negative = int(value.is_signed())
-        if self.exp_prec:  # Scientific notation
+        value = abs(value).normalize()
+
+        # Prepare scientific notation metadata.
+        if self.exp_prec:
```

Comment:
There's big chunks of scientific notation code interspersed with more "regular" rendering here – would you think the `exp_prec` bits could maybe be refactored into separate functions?